pkg/controllers/hpareplicassyncer/hpa_replicas_syncer_controller.go

```diff
@@ -0,0 +1,148 @@
+func (r *HPAReplicasSyncer) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling for HPA %s/%s", req.Namespace, req.Name)
+
+	hpa := &autoscalingv2.HorizontalPodAutoscaler{}
+	err := r.Client.Get(ctx, req.NamespacedName, hpa)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+
+		return controllerruntime.Result{}, err
+	}
+
+	workloadGR, scale, err := r.getGroupResourceAndScaleForWorkloadFromHPA(ctx, hpa)
+	if err != nil {
+		return controllerruntime.Result{}, err
+	}
+
+	err = r.updateScaleIfNeed(ctx, workloadGR, scale.DeepCopy(), hpa)
+	if err != nil {
+		return controllerruntime.Result{}, err
+	}
+
+	// TODO(@lxtywypc): Add finalizer for HPA and remove them
+	// when the HPA is deleting and the replicas have been synced.
+
+	return controllerruntime.Result{}, nil
+}
+
+// getGroupResourceAndScaleForWorkloadFromHPA parses GroupResource and get Scale
+// of the workload declared in spec.scaleTargetRef of HPA.
+func (r *HPAReplicasSyncer) getGroupResourceAndScaleForWorkloadFromHPA(ctx context.Context, hpa *autoscalingv2.HorizontalPodAutoscaler,
+) (schema.GroupResource, *autoscalingv1.Scale, error) {
+	gvk := schema.FromAPIVersionAndKind(hpa.Spec.ScaleTargetRef.APIVersion, hpa.Spec.ScaleTargetRef.Kind)
+	mapping, err := r.RESTMapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		klog.Errorf("Failed to get group resource for resource(kind=%s, %s/%s): %v",
+			hpa.Spec.ScaleTargetRef.Kind, hpa.Namespace, hpa.Spec.ScaleTargetRef.Name, err)
+
+		return schema.GroupResource{}, nil, err
+	}
+
+	gr := mapping.Resource.GroupResource()
+
+	scale, err := r.ScaleClient.Scales(hpa.Namespace).Get(ctx, gr, hpa.Spec.ScaleTargetRef.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If the scale of workload is not found, skip processing.
+			return gr, nil, nil
+		}
+
+		klog.Errorf("Failed to get scale for resource(kind=%s, %s/%s): %v",
+			hpa.Spec.ScaleTargetRef.Kind, hpa.Namespace, hpa.Spec.ScaleTargetRef.Name, err)
+
+		return schema.GroupResource{}, nil, err
+	}
+
+	return gr, scale, nil
+}
+
+// updateScaleIfNeed would update the scale of workload on fed-control plane
+// if the replicas declared in the workload on karmada-control-plane does not match
+// the actual replicas in member clusters effected by HPA.
+func (r *HPAReplicasSyncer) updateScaleIfNeed(ctx context.Context, workloadGR schema.GroupResource, scale *autoscalingv1.Scale, hpa *autoscalingv2.HorizontalPodAutoscaler) error {
+	// If the scale of workload is not found, skip processing.
+	if scale == nil {
+		klog.V(4).Infof("Scale of resource(kind=%s, %s/%s) not found, the resource might have been removed, skip",
+			hpa.Spec.ScaleTargetRef.Kind, hpa.Namespace, hpa.Spec.ScaleTargetRef.Name)
+
+		return nil
+	}
+
+	if scale.Spec.Replicas != hpa.Status.CurrentReplicas {
```

Comment:
Can you explain why you are using `CurrentReplicas` instead of `DesiredReplicas`?

Reply:
I think the `.status.currentReplicas` represents the real replicas of a managed workload, but the `.status.desiredReplicas` is the planned replicas which might not represent the current situation.

Reply:
I agree with you! Thanks!
I wrote it in the issue, and I wrote it wrong.
We really should use `DesiredReplicas`, which is the replicas described by spec in the member cluster:

https://github.com/kubernetes/kubernetes/blob/fd5f36e6a0e1d71d59e59620a4826eec6e6a6c7b/pkg/controller/podautoscaler/horizontal.go#L798

https://github.com/kubernetes/kubernetes/blob/fd5f36e6a0e1d71d59e59620a4826eec6e6a6c7b/pkg/controller/podautoscaler/horizontal.go#L907-L910

https://github.com/kubernetes/kubernetes/blob/fd5f36e6a0e1d71d59e59620a4826eec6e6a6c7b/pkg/controller/podautoscaler/horizontal.go#L1368-L1370

Reply:
Hi @lxtywypc, can you help update it?

Reply:
@chaunceyjiang 
I guess you mean the `.spec.replicas` represents `desired` replicas that match the definition of `.status.desiredReplicas` in HPA, right?

The difference between the two approaches is when to sync replicas. `after` and `before` replica change.  It makes more sense to sync `desired replicas`, it looks like the HPA in the Karmada control plane scales the workload :)

Reply:
Hi @lxtywypc How do you think?

Reply:
> I guess you mean the .spec.replicas represents desired replicas that match the definition of .status.desiredReplicas in HPA, right?


Yes. I mean exactly this.

Reply:
So I prefer to use `.status.desiredReplicas` because its meaning is consistent with `.spec.replicas`.

Reply:
Thanks @chaunceyjiang, it sounds good to me. 

Given following task relies on this controller, I suppose we can move this forward, and leave another task for this.
What do you think? @chaunceyjiang @XiShanYongYe-Chang @lxtywypc

Reply:
OK